src/order/well_founded_set.lean

```diff
@@ -147,11 +160,183 @@ end
 @[simp]
 theorem fintype.is_wf [fintype α] : s.is_wf := (finite.of_fintype s).is_wf
 
-@[simp]
-theorem is_wf_singleton : is_wf ({a} : set α) :=
-by simp [← finset.coe_singleton]
+end set
 
-theorem is_wf.insert (hs : is_wf s) : is_wf (insert a s) :=
-by { rw ← union_singleton, exact hs.union is_wf_singleton }
+lemma not_well_founded_swap_of_infinite_of_well_order_aux
+  {r : α → α → Prop} [is_well_order α r] {a : α} (h_inf : ({x : α | r a x}).infinite) :
+  {x : α | r ((is_well_order.wf : well_founded r).succ a) x}.infinite :=
+begin
+  have h_nonempty : ∃ x, r a x := h_inf.nonempty,
+  intro con,
+  apply h_inf,
+  convert con.insert ((is_well_order.wf : well_founded r).succ a),
+  ext x,
+  simp only [set.mem_insert_iff, set.mem_set_of_eq],
+  split; intro h,
+  { rcases is_trichotomous.trichotomous x ((is_well_order.wf : well_founded r).succ a) with c1 | c,
+    { exfalso,
+      cases (well_founded.lt_succ_iff h_nonempty x).1 c1 with hlt heq,
+      { exact is_asymm.asymm _ _ hlt h },
+      { rw heq at h,
+        have h_irr : is_irrefl α r := infer_instance,
+        have h_irr' := h_irr.irrefl,
+        exact h_irr' a h, } },
+    { exact c },
+    { apply_instance }, },
+  { rcases h with rfl | h,
+    { exact (is_well_order.wf : well_founded r).lt_succ h_nonempty },
+    { exact is_trans.trans _ _ _ ((is_well_order.wf : well_founded r).lt_succ h_nonempty) h } }
+end
+
+theorem not_well_founded_swap_of_infinite_of_well_order
+  [infinite α] {r : α → α → Prop} [is_well_order α r] :
+  ¬ well_founded (function.swap r) :=
+begin
+  haveI : is_strict_order α (function.swap r) := is_strict_order.swap _,
+  let m := (is_well_order.wf : well_founded r).min set.univ (set.univ_nonempty),
+  have hm : { x | r m x }.infinite,
+  { intro con,
+    have h : (set.univ : set α).infinite := set.infinite_univ,
+      apply h,
+      convert con.insert m,
+      ext x,
+      simp only [set.mem_insert_iff, true_iff, set.mem_univ, set.mem_set_of_eq],
+      rcases is_trichotomous.trichotomous x m with c1 | c,
+      { exfalso,
+        exact well_founded.not_lt_min _ set.univ _ (set.mem_univ x) c1 },
+      { exact c },
+      { apply_instance } },
+  rw [rel_embedding.well_founded_iff_no_descending_seq, not_not],
+  let f : ℕ → { a : α | ({x : α | r a x}).infinite },
+  { apply nat.rec,
+    { exact ⟨m, hm⟩ },
+    { exact λ _ x, ⟨_, not_well_founded_swap_of_infinite_of_well_order_aux x.2⟩ } },
+  refine ⟨rel_embedding.swap (rel_embedding.nat_lt (function.comp coe f) _)⟩,
+  intro n,
+  simp only [nat.rec_add_one, function.comp_app, subtype.coe_mk, subtype.val_eq_coe],
+  exact well_founded.lt_succ _ (set.infinite.nonempty (subtype.coe_prop (f n))),
+end
```

Comment:
It seems that's not how we proved B-W, and I haven't found that result. It's pretty much Ramsey's theorem for graphs on N. I use that same idea in #6226, so I should really try to isolate a common denominator lemma.

Reply:
I've provided a slightly primitive version of the monotone subsequence theorem you're looking for, and used it to simplify this proof. I'm not sure it's that much simpler overall, but we want that monotone subsequence theorem anyway, and it will already pay off when I fix #6226.